File: Python/Interview_problems/Splitwise_LLD.py
"""
SplitWise System /App
Functional Requirements:
    Create Users
    Create Groups
    Add Expenses in the system
    Split Types
        - Equal Split
        - Percentage Split
    Maintain balances
        Balance[A][B] --> status of B in A's Balance sheet
    Show balances of a User

    should  support payments
Non-Functional
1. Thread-Safety
2. Consistency across the balance sheet (ACID Compliance)
3. Extensibility to support multiple types of splits/expense


----- OUT of Scope -------
    Simplification feature of splitwise : Graph Algorithm using Priority Queue
    Settle

Core Entities:

User
Groups
Expense
SplitStrategy(Interface)
EqualSplit
PercentageSplit

SplitWiseService

BalanceManagerService
"""

import threading
from abc import ABC, abstractmethod
from typing import Dict, List


class User:
    def __init__(self, id: str, name: str):
        self.id = id
        self.name = name


class Split:
    def __init__(self, paid_by: User, paid_for: User, split_amount: float, expense_id: str):
        self.paid_by = paid_by
        self.paid_for = paid_for
        self.split_amount = split_amount
        self.expense_id = expense_id


class SplitStrategy(ABC):
    @abstractmethod
    def generate_split(self, expense: "Expense") -> List[Split]:
        raise NotImplementedError


class Expense:
    def __init__(self, id: str, paid_by: User, amount: float,
                 users_participated: List[User], strategy: SplitStrategy):
        self.id = id
        self.paid_by = paid_by
        self.amount = amount
        self.users_participated = users_participated
        self.strategy = strategy
        self.splits: List[Split] = []

    def calculate_share(self) -> List[Split]:
        self.splits = self.strategy.generate_split(self)
        return self.splits


# Splits the amount equally among every participant except the payer.
class EqualSplit(SplitStrategy):
    def generate_split(self, expense: Expense) -> List[Split]:
        splits: List[Split] = []
        if not expense.users_participated:
            return splits
        share = expense.amount / len(expense.users_participated)
        for user in expense.users_participated:
            if user.id == expense.paid_by.id:
                continue
            splits.append(Split(expense.paid_by, user, share, expense.id))
        return splits


# Splits the amount according to a percentage assigned to each participant.
class PercentageSplit(SplitStrategy):
    def __init__(self, percentages: Dict[str, float]):
        self.percentages = percentages

    def generate_split(self, expense: Expense) -> List[Split]:
        splits: List[Split] = []
        for user in expense.users_participated:
            if user.id == expense.paid_by.id:
                continue
            percentage = self.percentages.get(user.id, 0.0)
            splits.append(Split(expense.paid_by, user, expense.amount * percentage / 100.0, expense.id))
        return splits


class Group:
    def __init__(self, id: str, name: str):
        self.id = id
        self.name = name
        self.users: List[User] = []
        self.expenses: List[Expense] = []


# balances[a][b] is what user b owes user a.
class BalanceManagerService:
    def __init__(self):
        self._balances: Dict[str, Dict[str, float]] = {}
        self._lock = threading.Lock()

    def _ensure_user(self, user_id: str) -> None:
        self._balances.setdefault(user_id, {})

    def update_balance_sheet(self, expense: Expense, splits: List[Split]) -> None:
        with self._lock:
            for split in splits:
                payer_id = split.paid_by.id
                debtor_id = split.paid_for.id
                self._ensure_user(payer_id)
                self._ensure_user(debtor_id)

                self._balances[payer_id][debtor_id] = self._balances[payer_id].get(debtor_id, 0.0) + split.split_amount
                self._balances[debtor_id][payer_id] = self._balances[debtor_id].get(payer_id, 0.0) - split.split_amount

    def get_balance(self, user_id: str) -> Dict[str, float]:
        with self._lock:
            return dict(self._balances.get(user_id, {}))


class SplitWiseService:
    def __init__(self):
        self._users_repository: Dict[str, User] = {}
        self._group_repository: Dict[str, Group] = {}
        self._balance_manager_service = BalanceManagerService()

    def create_user(self, user_id: str, name: str) -> User:
        user = User(user_id, name)
        self._users_repository[user_id] = user
        return user

    def create_group(self, group_id: str, name: str, users: List[User]) -> Group:
        group = Group(group_id, name)
        group.users = users
        self._group_repository[group_id] = group
        return group

    def add_expense(self, expense_id: str, paid_by: User, amount: float,
                     users_participated: List[User], strategy: SplitStrategy) -> Expense:
        expense = Expense(expense_id, paid_by, amount, users_participated, strategy)
        splits = expense.calculate_share()
        self._balance_manager_service.update_balance_sheet(expense, splits)
        return expense

    def show_balance(self, user: User) -> None:
        balances = self._balance_manager_service.get_balance(user.id)
        any_non_zero = False
        for other_id, amount in balances.items():
            if amount == 0:
                continue
            any_non_zero = True
            if amount > 0:
                print(f"{other_id} owes {user.name} {amount:.2f}")
            else:
                print(f"{user.name} owes {other_id} {-amount:.2f}")
        if not any_non_zero:
            print(f"No pending balances for {user.name}")


def main() -> None:
    service = SplitWiseService()

    alice = service.create_user("u1", "Alice")
    bob = service.create_user("u2", "Bob")
    carol = service.create_user("u3", "Carol")

    service.create_group("g1", "Goa Trip", [alice, bob, carol])

    print("Equal split: Alice pays 300 for all three")
    service.add_expense("e1", alice, 300.0, [alice, bob, carol], EqualSplit())

    print("Percentage split: Bob pays 200, Alice 30%, Carol 70%")
    percentages = {"u1": 30.0, "u3": 70.0}
    service.add_expense("e2", bob, 200.0, [alice, bob, carol], PercentageSplit(percentages))

    for user in [alice, bob, carol]:
        print(f"\nBalances for {user.name}:")
        service.show_balance(user)


if __name__ == "__main__":
    main()

File: Golang/Interview_problems/Splitwise_LLD.go
/*
SplitWise System /App
Functional Requirements:
    Create Users
    Create Groups
    Add Expenses in the system
    Split Types
        - Equal Split
        - Percentage Split
    Maintain balances
        Balance[A][B] --> status of B in A's Balance sheet
    Show balances of a User

    should  support payments
Non-Functional
1. Thread-Safety
2. Consistency across the balance sheet (ACID Compliance)
3. Extensibility to support multiple types of splits/expense


----- OUT of Scope -------
    Simplification feature of splitwise : Graph Algorithm using Priority Queue
    Settle

Core Entities:

User
Groups
Expense
SplitStrategy(Interface)
EqualSplit
PercentageSplit

SplitWiseService

BalanceManagerService
*/

package main

import (
	"fmt"
	"sync"
)

type User struct {
	ID   string
	Name string
}

type Split struct {
	PaidBy      *User
	PaidFor     *User
	SplitAmount float64
	ExpenseID   string
}

type SplitStrategy interface {
	GenerateSplit(expense *Expense) []Split
}

type Expense struct {
	ID                string
	PaidBy            *User
	Amount            float64
	UsersParticipated []*User
	Strategy          SplitStrategy
	Splits            []Split
}

func (e *Expense) CalculateShare() []Split {
	e.Splits = e.Strategy.GenerateSplit(e)
	return e.Splits
}

// EqualSplit splits the amount equally among every participant except the payer.
type EqualSplit struct{}

func (EqualSplit) GenerateSplit(expense *Expense) []Split {
	var splits []Split
	if len(expense.UsersParticipated) == 0 {
		return splits
	}
	share := expense.Amount / float64(len(expense.UsersParticipated))
	for _, user := range expense.UsersParticipated {
		if user.ID == expense.PaidBy.ID {
			continue
		}
		splits = append(splits, Split{expense.PaidBy, user, share, expense.ID})
	}
	return splits
}

// PercentageSplit splits the amount according to a percentage assigned to each participant.
type PercentageSplit struct {
	Percentages map[string]float64
}

func (p PercentageSplit) GenerateSplit(expense *Expense) []Split {
	var splits []Split
	for _, user := range expense.UsersParticipated {
		if user.ID == expense.PaidBy.ID {
			continue
		}
		percentage := p.Percentages[user.ID]
		splits = append(splits, Split{expense.PaidBy, user, expense.Amount * percentage / 100.0, expense.ID})
	}
	return splits
}

type Group struct {
	ID       string
	Name     string
	Users    []*User
	Expenses []*Expense
}

// BalanceManagerService stores balances[a][b] = what user b owes user a.
type BalanceManagerService struct {
	mu       sync.Mutex
	balances map[string]map[string]float64
}

func NewBalanceManagerService() *BalanceManagerService {
	return &BalanceManagerService{balances: make(map[string]map[string]float64)}
}

func (b *BalanceManagerService) ensureUser(userID string) {
	if _, ok := b.balances[userID]; !ok {
		b.balances[userID] = make(map[string]float64)
	}
}

func (b *BalanceManagerService) UpdateBalanceSheet(_ *Expense, splits []Split) {
	b.mu.Lock()
	defer b.mu.Unlock()
	for _, split := range splits {
		payerID := split.PaidBy.ID
		debtorID := split.PaidFor.ID
		b.ensureUser(payerID)
		b.ensureUser(debtorID)

		b.balances[payerID][debtorID] += split.SplitAmount
		b.balances[debtorID][payerID] -= split.SplitAmount
	}
}

func (b *BalanceManagerService) GetBalance(userID string) map[string]float64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	result := make(map[string]float64)
	for k, v := range b.balances[userID] {
		result[k] = v
	}
	return result
}

type SplitWiseService struct {
	usersRepository       map[string]*User
	groupRepository       map[string]*Group
	balanceManagerService *BalanceManagerService
}

func NewSplitWiseService() *SplitWiseService {
	return &SplitWiseService{
		usersRepository:       make(map[string]*User),
		groupRepository:       make(map[string]*Group),
		balanceManagerService: NewBalanceManagerService(),
	}
}

func (s *SplitWiseService) CreateUser(userID, name string) *User {
	user := &User{ID: userID, Name: name}
	s.usersRepository[userID] = user
	return user
}

func (s *SplitWiseService) CreateGroup(groupID, name string, users []*User) *Group {
	group := &Group{ID: groupID, Name: name, Users: users}
	s.groupRepository[groupID] = group
	return group
}

func (s *SplitWiseService) AddExpense(expenseID string, paidBy *User, amount float64,
	usersParticipated []*User, strategy SplitStrategy) *Expense {
	expense := &Expense{ID: expenseID, PaidBy: paidBy, Amount: amount,
		UsersParticipated: usersParticipated, Strategy: strategy}
	splits := expense.CalculateShare()
	s.balanceManagerService.UpdateBalanceSheet(expense, splits)
	return expense
}

func (s *SplitWiseService) ShowBalance(user *User) {
	balances := s.balanceManagerService.GetBalance(user.ID)
	anyNonZero := false
	for otherID, amount := range balances {
		if amount == 0 {
			continue
		}
		anyNonZero = true
		if amount > 0 {
			fmt.Printf("%s owes %s %.2f\n", otherID, user.Name, amount)
		} else {
			fmt.Printf("%s owes %s %.2f\n", user.Name, otherID, -amount)
		}
	}
	if !anyNonZero {
		fmt.Printf("No pending balances for %s\n", user.Name)
	}
}

func main() {
	service := NewSplitWiseService()

	alice := service.CreateUser("u1", "Alice")
	bob := service.CreateUser("u2", "Bob")
	carol := service.CreateUser("u3", "Carol")

	service.CreateGroup("g1", "Goa Trip", []*User{alice, bob, carol})

	fmt.Println("Equal split: Alice pays 300 for all three")
	service.AddExpense("e1", alice, 300.0, []*User{alice, bob, carol}, EqualSplit{})

	fmt.Println("Percentage split: Bob pays 200, Alice 30%, Carol 70%")
	percentages := map[string]float64{"u1": 30.0, "u3": 70.0}
	service.AddExpense("e2", bob, 200.0, []*User{alice, bob, carol}, PercentageSplit{percentages})

	for _, user := range []*User{alice, bob, carol} {
		fmt.Printf("\nBalances for %s:\n", user.Name)
		service.ShowBalance(user)
	}
}
